emoji: Handle deleting of realm emojis.

HarshitOnGitHub opened the pull request:
Fixes: #4750.

timabbott commented:
@HarshitOnGitHub can you rebase to resolve the merge conflicts?

timabbott commented:
The code looks reasonable; the one thing I'm curious about is whether it's actually correct to s/emoji/emoijis in contexts where there are several of them.  @rishig is `emoji` the proper way to say a group of individual emoji, such that `emojis` isn't actually a valid word in English?

rishig commented:
Interesting question. I think emojis (though it does sound weird to me) is correct enough that it seems reasonable to use it where it would increase clarity of the code. Associated press it seems has settled on emojis, citing "tsunamis" as the pattern they're following, rather than "sushi".

timabbott commented:
@HarshitOnGitHub thanks for working on this!  There are a few more changes this needs:
* Can you fix the lint/test failures?  
* Remove the API change for `realm_emojis` for now.  Maybe while we're at it, remove the `get_realm_emoji_uncached` change so that it stays consistent?  Those two changes sorta belong together.
* Is there a purpose in sending the deactivated realm emoji to the frontend at all?  I guess maybe we need them to render existing emoji reactions correctly?  I sorta feel like we will want to do further restructing of how reactions to realm emoji work so that we can support creating a new emoji with the same name as a deactivated emoji.  But that can go into the reactions PR.

I think we can merge this after addressing the first two items there.

HarshitOnGitHub commented:
@timabbott I have fixed the tests and made some changes that were pending so this is now ready. For now I have dropped the "emoji -> emojis" commits and will deal with them in a different issue. I will deal with the issue of allowing users to add a realm emoji with same name as a deactivated emoji in a separate PR as it will require more complex changes.

timabbott commented:
@HarshitOnGitHub this looks great!  After renumbering the migration, I merged the first 3 commits.  I have a concern that I'd like you to address on the 4th, which is that I think the better solution is for `emoji.js` to export an `active_realm_emojis` object that has the deactivated emoji excluded, and then to avoid having other modules access `exports.realm_emojis` directly (with a comment above the definitions explaining which is to be used for what).  Might want to rename `exports.realm_emojis` to something like `exports.all_realm_emojis` as well to avoid confusion.  

I think we only will need the full `all_realm_emojis` in `reactions.js`; does that fit your understanding?

HarshitOnGitHub commented:
@timabbott Yeah I had similar plans but left in the current state to avoid having merge conflicts in other branches. I had planned to do the refactoring once other branches go in. I will make the changes related to realm emojis in this branch! :slightly_smiling_face:

HarshitOnGitHub commented:
@timabbott Just FYI, I have updated this PR, have a look whenever you get some time. :)

timabbott commented:
@HarshitOnGitHub nice, I merged the first 2 commits.  Posting a few comments on the next ones.

timabbott commented on static/js/emoji_picker.js:
```diff
@@ -69,13 +69,15 @@ function generate_emoji_picker_content(id) {
+                    is_realm_emoji: emojis[emoji_name].is_realm_emoji,
+                    url: emojis[emoji_name].url,
+                };
+            }
```
I don't understand this change; it would seem to me that it'll remove all the non-realm-emoji from this list?

As a sidenote, do you think we want to show in the picker any realm emoji that you've reacted to already, even if they are deactivated (so that you can unreact)?

HarshitOnGitHub replied:
>I don't understand this change; it would seem to me that it'll remove all the non-realm-emoji from this list?

This clearly seems to be a mistake on my side, will fix it!

>As a sidenote, do you think we want to show in the picker any realm emoji that you've reacted to already, even if they are deactivated (so that you can unreact)?

I think it will be good to not include deactivated realm emojis in the emoji picker since it can lead to confusion in case of multiple emojis with same name. If an user wants to unreact to an deactivated they have the option of doing so directly via clicking on that realm emoji which seems to be much more cleaner to me.

timabbott replied:
Yeah, OK, that makes sense to me.

timabbott commented:
> reactions: Don't send add/remove requests for deactivated realm emojis.

On this change, presumably we do want to allow unreacting to deactivated realm emojis, since otherwise those are stuck in the UI forever.  @rishig @HarshitOnGitHub do you agree?

So I think we want some minor tweaks there.

Otherwise, this is looking great @HarshitOnGitHub!

HarshitOnGitHub commented:
>presumably we do want to allow unreacting to deactivated realm emojis

I think that will be possible only when we change the realm emojis to be addressed by `id` instead of `name`, so I thought it would be good to include this change for now until we finish off name -> id migration for realm emojis.

timabbott commented:
OK, that make sense to me.  We can probably merge with that piece of the puzzle still open (though that's probably a good thing to get started on, since it's mostly orthogonal to other projects, right?).  Or have you already started working on it in another PR?

So we should be good once we resolve the potential bug I mentioned above.

HarshitOnGitHub commented:
@timabbott Pushed an updated version! :)

timabbott commented:
Merged after adding a comment explaining the `emoji_picker.js` change. 

Thanks @HarshitOnGitHub!  We do still have a follow-up issue to resolve with unreacting to deactivated realm emoji, but I think we can pursue that in another PR.  Can you open an issue though, so we don't forget to do it?  

I guess maybe there's also a follow-up issue to update the realm emoji settings page to show deactivated realm emoji and let you reactivate them?  

The final thing to think about is whether we'll want to do a migration to how we store realm emoji to store the uploaded files primarily by ID rather than name?  

Anyway, those are some things to think about for continuing this effort.
